profiles/repositories.py

```diff
@@ -0,0 +1,59 @@
+import collections
+import string
+from abc import ABC, abstractmethod
+
+from flask_sqlalchemy import SQLAlchemy
+from sqlalchemy.orm.exc import NoResultFound
+
+from profiles.exceptions import ProfileNotFound
+from profiles.models import ID_LENGTH, Profile
+from profiles.utilities import generate_random_string
+
+
+class Profiles(ABC, collections.Sized):
+    @abstractmethod
+    def add(self, profile: Profile) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def get(self, profile_id: str) -> Profile:
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_by_orcid(self, orcid: str) -> Profile:
+        raise NotImplementedError
+
+    @abstractmethod
+    def next_id(self) -> str:
+        raise NotImplementedError
+
+
+class SQLAlchemyProfiles(Profiles):
+    def __init__(self, db: SQLAlchemy) -> None:
+        self.db = db
+
+    def add(self, profile: Profile) -> None:
+        self.db.session.add(profile)
+
+    def get(self, profile_id: str) -> Profile:
+        try:
+            return self.db.session.query(Profile).filter_by(id=profile_id).one()
+        except NoResultFound as exception:
+            raise ProfileNotFound('Profile with the ID {} not found'.format(profile_id)) \
+                from exception
+
+    def get_by_orcid(self, orcid: str) -> Profile:
+        try:
+            return self.db.session.query(Profile).filter_by(orcid=orcid).one()
+        except NoResultFound as exception:
+            raise ProfileNotFound('Profile with the ORCID {} not found'.format(orcid)) \
+                from exception
+
+    def next_id(self) -> str:
+        while True:
```

Comment:
perhaps limiting the number of iterations and throwing an exception in case the limit is hit will prevent bugs in the filter from escalating to an infinite loop

Reply:
Probably a good idea. Timeouts should normally kick in, but no need to get that far.